Add overview, how it works, and glossary docs (#15)
divviup/public-docs: docs/how-it-works.mdx

---
sidebar_position: 2
---

# How it Works

Learn the high-level workings of Divvi Up and how it accomplishes the
privacy-preserving property.

## Breakdown

Divvi Up functionality can be broadly broken down into five steps.

See the [Distributed Aggregation Protocol (DAP)][DAP] for a highly detailed
specification of the protocol that underpins Divvi Up.

<!-- TODO(inahga): Consider making this a more detailed `mermaid` diagram -->

![](../static/how-it-works.png#gh-dark-mode-only)
![](../static/how-it-works-light.jpg#gh-light-mode-only)

> ### Step 1: Client generate measurements

Reviewer: ```suggestion
### Step 1: Client generates measurements
```

An app developer wants to collect statistics about their app, such as user
behavior, demographics, or system information. The developer writes app
functionality that takes measurements of these properties.

> In Divvi Up, the developer creates a Task, which encodes data about the

Reviewer: This is included in step 1 but feels like step 0, as it happens before `Client generates measurements`

Author: Let me see how a step 0 looks--this might also be a handy place to include a quick blurb on the state of onboarding (i.e. get invited by existing user or ask us).

measurement being taken. This encoded data is stored in the app and used by
client libraries. The data includes the measurement type, the two Divvi Up
aggregators being used, the identifier for the task, and how aggregates are to
be queried. The measurement type includes, but is not limited to, counts, sums,
and histograms.

For each measurement, the subsequent steps are performed.

### Step 2: Client divides the measurement

> Using a client library, the app divides the measurement into two shares, known

Reviewer: I know this is a hard task and we may have to make several iterative passes at it, but we should make some passing effort at explaining the nature of "dividing" the measurements. We may even want to call it sharding because dividing a scalar has a clear and incorrect meaning here. It's even less intuitive to a naïve reader to consider (by any implicit definition of sharding/splitting/dividing) "splitting a boolean in half"

Author: I landed on `s/divides/randomly shards/`.

as report shares. One report share is meant for a Leader aggregator, and a
Helper aggregator. Each share is encrypted with the respective aggregator's

> public key.

Reviewer: As a naïve reader: "Where do I get this key?"

It's probably worth saying the DAP client knows where to get the key from the internet and will do so automatically

> ### Step 3: Client sends to the aggregators

Reviewer: ```suggestion
### Step 3: Client sends reports to the aggregators
```

Or "shares"

The client library sends each report share via HTTPS to the leader and helper
aggregator.

The aggregators in use are determined by the app developer and encoded in the
task data. Each aggregator is operated independently by two different
organizations. Often, Divvi Up will act as the leader, while another
organization operates as a helper.

<!-- TODO(#11): link BYOH guide here -->

Crucially, these organizations are non-colluding. They do not directly share
report shares or otherwise conspire to defeat the protocol. If you do not trust
any of the already-available third-party organizations that Divvi Up partners
with, you can self-host your own helper aggregator.

> At this stage, aggregators can also reject out-of-range or invalid measurements

Reviewer: ```suggestion
At this stage, aggregators also reject out-of-range or invalid measurements
```

My understanding is that this verification step is always performed; is that wrong?

Author: That's correct. I should probably also clarify that the range is determined by the task parameters.

### Step 5: Subscriber collects aggregate metrics

The subscriber operates a collector, which queries each aggregator to collect
the aggregate shares. Depending on the query type, this could be a query over a
time interval, or for an arbitrary batch of reports.

Aggregates often aren't available immediately, depending on the query type. The
collector will poll the aggregator until it's ready.

Each aggregate share is encrypted to the collector. The collector reconstitutes
the aggregate shares, yielding the aggregate result and the number of reports
included in the result.

<!--
TODO(https://github.com/divviup/janus-ops/issues/1005): we should make mention of the adapters we
come up with here.
-->

Off-the-shelf collectors that more elegantly handle the collection process and
presentation of data are currently in development.

:::note 
Similar to with report shares, a collector does not directly talk to
both aggregators. Instead the helper's aggregate share is relayed through the
leader. Because each share is encrypted to the collector, the leader cannot read
the aggregate shares or reconstitute the aggregate.
:::

> [DAP]: https://datatracker.ietf.org/doc/draft-irtf-cfrg-vdaf/

Reviewer: This is vdaf not dap

---
sidebar_position: 2
---

# How it Works

> Learn the high-level workings of Divvi Up and how it accomplishes the

Author: I ended up going with this

> # How It Works
> 
> Learn how Divvi Up works and how it protects privacy.
> 
> ## Breakdown
> 
> Divvi Up ingests measurements and aggregates them into metrics, while protecting the privacy of each individual measurement. Its functionality can be broadly broken down into several steps.

I prefer the terse, active voiced headline--note that docusaurus implicitly uses this as the page description, but I was otherwise missing a short description of what precisely Divvi Up does.

> Learn the high-level workings of Divvi Up and how it accomplishes the
> privacy-preserving property.

Reviewer 2: Rewriting for conciseness and active voice
```suggestion
Learn how Divvi Up works and how it protects privacy.
```

## Breakdown

Divvi Up functionality can be broadly broken down into several steps.

See the [Distributed Aggregation Protocol (DAP)][DAP] for a highly detailed
specification of the protocol that underpins Divvi Up.

> <!-- TODO(inahga): Consider making this a more detailed `mermaid` diagram -->

Author: It looks like .mdx doesn't actually support HTML comments https://github.com/mdx-js/mdx/issues/2047, so this is a syntax error (even though it works :shrug:).

The correct syntax is `{/* foobar */}` (ew).

![](../static/how-it-works.png#gh-dark-mode-only)
![](../static/how-it-works-light.jpg#gh-light-mode-only)

### Step 0: Subscriber sets up Divvi Up

> Subscribers are entities, such as app developers, who wish to collect quantifiable metrics
> about their end-user applications.

Reviewer 2: ```suggestion
Subscribers are organizations, such as app developers, who wish to collect quantitative metrics
about their end-user applications.
```

> The subscriber chooses a Distributed Aggregation Protocol (DAP) client library, which is used for interacting with any DAP-compatible metrics aggregators, such as the ones provided by Divvi Up and its partners.
> A client library for [Rust (janus_client)][janus_client] and [TypeScript (divviup-ts)][divviup-ts] is available.

Reviewer 2: ```suggestion
The subscriber chooses a Distributed Aggregation Protocol (DAP) client library, which is used for interacting with any DAP-compatible metrics aggregators, such as those provided by Divvi Up and its partners.
Client libraries for [Rust (janus_client)][janus_client] and [TypeScript (divviup-ts)][divviup-ts] are available.
```

> For each metric, the subscriber creates a task, whose parameters encode data about the measurement being taken.

Reviewer 2: ```suggestion
For each metric, the subscriber creates a task, which encodes metadata about how the measurement will be made. The task's parameters are shared with both aggregators.
```

### Step 1: Client generates measurements

The subscriber writes and deploys app functionality that takes measurements corresponding to each task.

> Using a DAP client library, they also encode some of the task parameters into the app, such as the task identifier, measurement type, and the Divvi Up aggregators being used.

Reviewer 2: ```suggestion
They invoke a DAP client library, and provide some of the task parameters to it, such as the task identifier, measurement type, and the Divvi Up aggregators being used.
```

For each measurement, the subsequent steps are performed.

### Step 2: Client shards the measurement

> With the client library, the app randomly shards the measurement into two shares, known

Author: I landed on
> The app uses the DAP client library to randomly shard the measurement into two
parts, known as report shares. One report share is for the leader aggregator,
and the other for the helper aggregator.

I do think the `randomly` part is important, as it conveys that the report isn't naively bitwise sharded or scalar divided. Perhaps the mechanism of sharding justifies its own sentence?

> With the client library, the app randomly shards the measurement into two shares, known
> as report shares. One report share is for a leader aggregator, and the other for a
> helper aggregator.

Reviewer 2: ```suggestion
The app uses the DAP client library to randomly shard the measurement into two shares, known
as report shares. One report share is for a leader aggregator, and the other for a
helper aggregator.
```

> Both aggregators advertise a public key and possess its corresponding private key.

Reviewer 2: ```suggestion
Each aggregator advertises a public key corresponding to its private key.
```

The client library encrypts each share with the respective aggregator's public key.

### Step 3: Client sends shares to the aggregators

The client library sends each report share via HTTPS to the leader and helper
aggregator.

Aggregators reject out-of-range or invalid measurements, as defined in the task parameters. This is done without knowledge of the full input measurement.

The aggregators in use are determined by the app developer and encoded in the

> task parameters. Each aggregator is operated independently by two different

Author: I landed on
> Each aggregator is operated by different organizations.

Short and sweet, and I think the next sentence conveys that it's leader -> divviup, helper -> someone else?

> The aggregators in use are determined by the app developer and encoded in the
> task parameters. Each aggregator is operated independently by two different
> organizations. Often, Divvi Up will act as the leader, while another
> organization operates as a helper.

Reviewer 2: ```suggestion
The app developer determines which aggregators are used, and encodes their choices in the
task parameters. The two aggregator are operated independently by two different
organizations. Often, Divvi Up will act as the leader, while another
organization operates as a helper.
```

Author: I ended up striking this sentence, as it was redundant information--we established that the subscriber chooses the aggregators further up.

<!-- TODO(#11): link BYOH guide here -->

> Crucially, these organizations are non-colluding. They do not directly share
> report shares or otherwise conspire to defeat the protocol. If you do not trust
> any of the already-available third-party organizations that Divvi Up partners
> with, you can self-host your own helper aggregator.

Reviewer 2: ```suggestion
Crucially, these organizations are non-colluding. They do not directly share
report shares or otherwise conspire to defeat the protocol. If you do wish to use
any of the already-available third-party organizations that Divvi Up partners
with, you can self-host your own helper aggregator.
```

:::note
Both shares are actually sent to the leader aggregator. The leader
aggregator relays the helper's report share to the helper. Because the helper's
report share is encrypted to the helper, the leader cannot read it and the
non-collusion property is preserved.

> This is done to simplify client library implementation and reduce the number of
> client network round-trips.

Reviewer 2: Non-split uploads also make life easier for the helper, by pushing more of the coordination work onto the leader.
```suggestion
This is done both to simplify client library implementation, reduce the number of
client network round-trips, and reduce the operational requirements of the helper
aggregator.
```

:::

### Step 4: Aggregators perform aggregation

> The aggregators use [multi-party computation][MPC] to work together to compute

Reviewer 3: ```suggestion
The aggregators work together to compute
```

(IMO, move the mention of MPC to the next paragraph -- it's a useful detail but this sentence should summarize what the aggregators are accomplishing without adding too much detail.)

an aggregate across a batch of report shares, without revealing the report share

> to each other and without having any knowledge of the actual submitted

Reviewer 3: ```suggestion
to each other and without having any knowledge of the originally-submitted
```

(word choice)

measurement.

> The protocol and cryptography involved is described in [Verifiable Distributed

Reviewer 3: IMO, mention & link to DAP as well -- DAP defines the protocol, VDAF defines the cryptography.

Author: I mentioned DAP but didn't link it, as it's already linked above.

Aggregation Functions (VDAF)][VDAF].

> Computation will yield two aggregate shares, one for each aggregator.

Reviewer 2: nit: clarity
```suggestion
Computation will yield two aggregate shares, one from each aggregator.
```

> Computation will yield two aggregate shares, one for each aggregator.

Reviewer 3: ```suggestion
At the end of the aggregation process, each aggregator will have computed a share of the overall aggregate. Neither share reveals any information about the final value of the aggregate, but the shares can be combined to determine the aggregate value. Each aggregator stores its aggregate share until the next step.
```

Author: I went with
> At the end of the process, each aggregator will have computed a share of the
overall aggregate. Neither share reveals information about the final aggregate.
>
> Each aggregator stores its aggregate share until the next step.

I omitted the combining part, as that's somewhat redundant with the later paragraph about the collector combining them.

### Step 5: Subscriber collects aggregate metrics

The subscriber operates a collector, which queries each aggregator to collect
the aggregate shares. Depending on the query type, this could be a query over a
time interval, or for an arbitrary batch of reports.

> Aggregates often aren't available immediately, depending on the query type. The

Reviewer 3: ```suggestion
Aggregates often aren't available immediately, depending on the VDAF in use. The
```

Author: I went with "measurement type" as this would otherwise be the first mention of VDAF and require explanation.

collector will poll the aggregator until it's ready.

Each aggregate share is encrypted to the collector. The collector reconstitutes
the aggregate shares, yielding the aggregate result and the number of reports

> included in the result.

Reviewer 3: We also get the time interval spanned by the reports; might as well mention it.

<!--
TODO(https://github.com/divviup/janus-ops/issues/1005): we should make mention of the adapters we
come up with here.
-->

Off-the-shelf collectors that more elegantly handle the collection process and
presentation of data are currently in development.

:::note

> Similar to with report shares, a collector does not directly talk to

Reviewer 3: ```suggestion
Similar to the upload process, the collector does not directly talk to
```

> Similar to with report shares, a collector does not directly talk to
> both aggregators. Instead the helper's aggregate share is relayed through the
> leader. Because each share is encrypted to the collector, the leader cannot read
> the aggregate shares or reconstitute the aggregate.

Reviewer 2: ```suggestion
As with report shares, a collector does not directly talk to
both aggregators. Instead the helper's aggregate share is relayed through the
leader. Because each share is encrypted to the collector, the leader cannot read
the aggregate shares or reconstitute the aggregate.
```
